### PR TITLE
Convert student instance question page to async

### PR DESCRIPTION
**Pull request by jonatanschroeder**
Also:

* Converted sproc only used in this context to JS/TS;
* Converted CJS to ESM;
* Moved a "shared" library only used in this context to the same router file.

### CODE REVIEW

File: apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js [hunk 1]
```diff
@@ -1,41 +1,169 @@
...
+ */
+async function getValidVariantId(req, res) {
+  return (await getValidVariant(req.body.__variant_id, res.locals.instance_question.id)).id;
+}
+
+async function processFileUpload(req, res) {
+  if (!res.locals.assessment_instance.open) throw new Error(`Assessment is not open`);
+  if (!res.locals.authz_result.active) {
+    throw new Error(`This assessment is not accepting submissions at this time.`);
+  }
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js:65]: Can we use non-500 status codes here? Ditto for these other `process...` functions.

**Review by nwalters512 — APPROVED** [review]
Won't block on this comment since it's not a regression, but improved HTTP statuses would be nice to have!

File: apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js (hunk 1 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js]: Done in 17500451b78b79451bb559cfa3b45c4952ce821a.

**Comment by jonatanschroeder** [comment]
@nwalters512 there were a few changes after your review that I'd feel more comfortable if you could take a double-check.

File: apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js [hunk 2]
```diff
@@ -68,173 +199,105 @@ function processSubmission(req, res, callback) {
...
+      const requireOpen = true;
+      const closeExam = true;
+      const overrideGradeRate = false;
+      await gradeAssessmentInstanceAsync(
+        res.locals.assessment_instance.id,
+        res.locals.authn_user.user_id,
+        requireOpen,
+        closeExam,
+        overrideGradeRate,
+        res.locals.client_fingerprint_id,
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js:266]: 👀 was this a bug before?

**Review by nwalters512 — APPROVED** [review]

File: apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js (hunk 2 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js:266]: It was indeed missing, so a client fingerprint would not be captured in a time-finish event that happened when the student was in an instance question.

**Pull request merged**